All new features

muhammad-di opened the pull request.

muhammad-di committed:
refactor: in a FilmDbStorage class updateLikes, updateGenre methods

SvetIbr committed:
refactor: add RecommendationsService

SvetIbr committed:
fix:formatted

SvetIbr committed:
refactor: add endpoint for delete user or film by id

SvetIbr committed:
fix: change to lambda in FilmDbStorage (method getIdLikedFilmsByUser)

SvetIbr committed:
Merge pull request #42 from muhammad-di/add-recommendations

refactor: add RecommendationsService

SvetIbr committed:
fix: formatted

SvetIbr committed:
fix: formatted

SvetIbr committed:
Merge pull request #44 from muhammad-di/add-remove-endpoint

Add remove endpoint

roman3350 committed:
refactor: add filtering by genre and year in the output of the most popular movies

roman3350 committed:
refactor: add filtering by genre and year in the output of the most popular movies

roman3350 committed:
Merge remote-tracking branch 'origin/add-most-populars' into add-most-populars

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/storage/film/dao/impl/FilmDbStorage.java

roman3350 committed:
Merge pull request #45 from muhammad-di/add-most-populars

Добавил в метод getMostPopularFilms в FilmDbStorage возможность фильтрации по жанру и году. Создал класс Director. Добавил поле directors в классе Film.

maershov committed:
+film controller for search

maershov committed:
*changed query condition from and to or
+order by likes

maershov committed:
*style

maershov committed:
*style

maershov committed:
+db method

maershov committed:
+common films controller

maershov committed:
controller checks deleted

muhammad-di committed:
feat: added classes Director, DirectorDbStorage, DirectorService, DirectorController

muhammad-di committed:
Merge branch 'feature-film-update' into add-director

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java

maershov committed:
Merge pull request #47 from muhammad-di/add-common-films

Add common films

tarasovEgor committed:
feat: reviews added

tarasovEgor committed:
feat: code style added

muhammad-di committed:
feat: Director fitch implemented

tarasovEgor committed:
feat: create review method refactored, additional checks added to update/get/delete review by id

muhammad-di committed:
Merge branch 'develop' into add-director

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/model/Director.java
#	src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/film/InMemoryFilmStorage.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/film/dao/FilmStorage.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/film/dao/impl/FilmDbStorage.java

muhammad-di committed:
feat: Director fitch implemented

muhammad-di committed:
feat: Director fitch implemented

tarasovEgor committed:
Merge pull request #48 from muhammad-di/add-reviews

Add reviews

muhammad-di committed:
Merge branch 'develop' into add-director

# Conflicts:
#	src/main/resources/schema.sql

muhammad-di committed:
feat: Director fitch implemented

maershov committed:
Merge branch 'develop' into add-search

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/storage/film/InMemoryFilmStorage.java

maershov committed:
Merge pull request #50 from muhammad-di/add-director

Add director

maershov committed:
Merge branch 'develop' into add-search

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/storage/film/InMemoryFilmStorage.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/film/dao/impl/FilmDbStorage.java

maershov committed:
storage film changed likes fields

maershov committed:
style

maershov committed:
Merge pull request #46 from muhammad-di/add-search

Add search

muhammad-di committed:
feat: Feed fitch implemented

muhammad-di committed:
refactor:  sql query of getCommonFilms method

muhammad-di committed:
refactor: schema and DB diagram

muhammad-di committed:
refactor: README.md reference on a diagram file

muhammad-di committed:
Merge pull request #52 from muhammad-di/add-feed

Add feed

muhammad-di committed:
refactor: UserDbStorage.java

EugeneLenkevich commented on src/main/java/ru/yandex/practicum/filmorate/service/DirectorService.java:
```diff
@@ -0,0 +1,59 @@
+    }
+
+    public Director getDirectorById(Long id) throws DirectorDoesNotExistException {
+        if (!storage.containsDirector(id)) {
```
!storage.containsDirector(id)

Вот тут мы сначала проверяем на наличие записи а потом ее достаем

storage.getDirectorById(id)

А можно это же сделать за одну операцию: вытащить запись по id, но если она null, то выкинуть исключение, один запрос к БД вместо двух :)

muhammad-di replied:
Я если честно здесь был в метаниях))
Хотел сделать более читабельный код
Но когда ты сказал, что мы теряем в продуктивности, всё стало очевидным))

EugeneLenkevich commented on src/main/java/ru/yandex/practicum/filmorate/service/DirectorService.java:
```diff
@@ -0,0 +1,59 @@
+        return storage.getDirectorById(id);
+    }
+
+    public Director createDirector(Director director) throws DirectorAlreadyExistException {
```
Класс называется DirectorService, методы принимают и возвращают объекты Director, наверное из имен можно убрать упоминание имени объекта, может будет покороче без потери выразительности :)

directorService.create(...)
directorService.update(...)
directorService.findById(...)

EugeneLenkevich commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -47,22 +75,68 @@ public Film update(Film film) throws FilmDoesNotExistException, InvalidFilmPrope
+            throw new DirectorDoesNotExistException(message);
+        }
+        if (sort.equals(LIKES_SORT)) {
+            return storage.getFilmsWithDirectorIdSortedByLikes(directorId);
```
storage.getFilmsWithDirectorIdSortedByLikes(directorId)

А что если завести только один метод и передавать в него сортировку. Все таки сортировка - это обязанности хранилища ну и не надо будет под каждую сортировку иметь свой метод :)

EugeneLenkevich commented on src/main/java/ru/yandex/practicum/filmorate/service/RecommendationsService.java:
```diff
@@ -0,0 +1,55 @@
+        if (maxCount > 0) {
+            idUsersBySimilarLikes.sort(Comparator.naturalOrder());
+            return idUsersBySimilarLikes.stream()
+                    .flatMap(idUser -> filmStorage.getIdLikedFilmsByUser(idUser).stream())
```
.flatMap(idUser -> filmStorage.getIdLikedFilmsByUser(idUser).stream())

Вот тут в цикле вызывается метод хранилища, то есть каждый раз делается запрос к БД. Запрос к БД - это дорогая операция, это может быть не особо оптимально, а получится достать все необходимые записи за меньшее количество запросов? :)

muhammad-di replied:
Мы постараемся реализовать менее ресурсный запрос)

EugeneLenkevich replied:
Самый простой вариант в запросе там где
where id = ...
сделать
where id in (...)

EugeneLenkevich commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -55,26 +55,49 @@ public User update(User user) throws UserDoesNotExistException, InvalidUserPrope
+        return storage.getFeedOfUser(id);
+    }
+
+    private void containsUser(Long id) throws UserDoesNotExistException {
```
Не уверен что для метода выбрано самое выразительное наименования, судя по названию оно проверяет наличие пользователя, однако еще и выкидывает исключение при этом. 
Может вообще вот так делать :)

if(!containsUser(id)) {
  throw new UserDoesNotExistException(...);
}

muhammad-di replied:
А что если назвать метод "findEventByUserId" ? Я сперва подумал о названии "findEventById", но боюсь из названия не очевидно кому принадлежит Id, Event или User :)

EugeneLenkevich replied:
Тут уже другое, это разные поля, id и user_id

EugeneLenkevich commented on src/main/java/ru/yandex/practicum/filmorate/storage/review/dao/impl/ReviewDbStorage.java:
```diff
@@ -0,0 +1,345 @@
+
+        String getAllReviewsByFilmId = "SELECT * " +
+                "                       FROM reviews " +
+                "                       WHERE film_id = " + filmId +
```
" + filmId +

В тексте запросов все же лучше плейсхолдер, то есть параметр использовать, а не конкатенацию строк. Для чисел это еще достаточно безопасно, но вот в случае строк это уязвимость, злоумышленник может указать строку что то типа "; drop users; --" :)

EugeneLenkevich commented on src/main/resources/data.sql:
```diff
@@ -1,71 +1,43 @@
+--MERGE INTO director KEY (director_id) values (5, 'Nanine Hargrove');
 
+-- Films ---------------------------------------------------------------------------------------------------------------
+--INSERT INTO film (name, release_date, duration, description, mpa) VALUES ('S.F.W.', '2002-10-13', 136, 'Donec dapibus. Duis at velit eu est congue elementum.', 3);
```
Закомментированный код менее читаем и понятен для других разработчиков, что может привести к ошибкам и проблемам при компиляции или выполнении программы. Поэтому рекомендуется всегда удалять закомментированный код из своего кода программы.

Ну и, скорее всего, все вокруг поменяется и он станет не актуальным, а для истории есть git :)

EugeneLenkevich commented on src/main/resources/schema.sql:
```diff
@@ -11,46 +32,107 @@ CREATE TABLE IF NOT EXISTS film
+
+CREATE TABLE IF NOT EXISTS reviews (
+    review_id   INTEGER GENERATED BY DEFAULT AS IDENTITY PRIMARY KEY,
+    content     VARCHAR(250),
```
250 символов на описание, а не маловато или нормально? :)

muhammad-di replied:
Я долго не мог решить сколь же символов взять,  посетил сайт imdb и выбрал от туда)
А что если взять 700 символов или это много?

EugeneLenkevich replied:
Я бы вот так сделал, текстовый блоб
content     VARCHAR,

muhammad-di replied:
А как ты думаешь, Евгений, может нам и для описания review тоже связь VARCHAR?

muhammad-di replied:
Спасибо, за ревью, Евгений)
Как всегда полно полезной информации)

EugeneLenkevich replied:
> А как ты думаешь, Евгений, может нам и для описания review тоже связь VARCHAR?

Может быть, насколько понимаю в review может быть большой кусок текста

EugeneLenkevich reviewed (changes requested):
Привет, Мухаммад! Спасибо за PR.

Отличная работа, спасибо :)
Есть совсем небольшие пометки, посмотри комментарии ниже, пожалуйста

muhammad-di committed:
refactor: everything

muhammad-di committed:
refactor: schema.sql

muhammad-di committed:
style: all project

SvetIbr committed:
fix: formatted RecommendationService

SvetIbr committed:
fix: formatted

SvetIbr committed:
fix: formatted

SvetIbr committed:
fix: formatted

SvetIbr committed:
fix: formatted

SvetIbr committed:
Merge pull request #54 from muhammad-di/add-recommendations

Add recommendations

muhammad-di committed:
refactor: User related classes

muhammad-di committed:
refactor: findById method in ReviewDbStorage

muhammad-di committed:
refactor:  findById method in FilmDbStorage and FilmService

muhammad-di committed:
refactor:  findById method in FilmDbStorage and FilmService
                           in GenreDbStorage and GenreService
                           in MpaDbStorage and MpaService
                           in UserDbStorage and UserService

muhammad-di committed:
refactor: findById method in FilmDbStorage and FilmService

EugeneLenkevich reviewed (approved):
Теперь все отлично, спасибо :)
